android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt

```diff
@@ -143,12 +149,20 @@ class TranslationsMiddleware(
      * Checks if the translations engine supports the device architecture and updates the state.
      *
      * @param context Context to use to dispatch to the store.
+     * @param initialization The initialization the translations engine should complete on a
+     * successful check that the engine is supported.
      */
     private fun requestEngineSupport(
         context: MiddlewareContext<BrowserState, BrowserAction>,
+        initialization: () -> Unit,
     ) {
         engine.isTranslationsEngineSupported(
             onSuccess = { isEngineSupported ->
+                if (isEngineSupported) {
+                    scope.launch {
+                        initialization.invoke()
+                    }
+                }
```

Comment:
@MatthewTighe Could I get your thoughts on this? We basically have 2 async functions where the latter depends on the result of the former. I feel we could also do this by dispatching another action and then just hitting the middleware again.

Reply:
Just to mention, in case it changes the best choice, this will grow to include initialization of:
* Languages (this patch)
* Global Settings for Language Preferences #5726
* Global Setting for "Never translate this site" (Currently on session, realized it should be refactored to browser.)
* Model Download States #5624

Reply:
(Just spotted  a [ test failure](https://firefoxci.taskcluster-artifacts.net/YFsPmM4-QAijMLuPgveFDA/0/public/reports/tests/testReleaseUnitTest/classes/mozilla.components.browser.state.engine.middleware.TranslationsMiddlewareTest.html#WHEN%20InitAction%20is%20dispatched%20AND%20succeeds%20THEN%20SetSupportedLanguagesAction%20is%20dispatched), probably related to how this is invoked. Will wait to fix, pending this conversation. Dispatching again might make testing easier, as a consideration.)

Reply:
Thanks for taking a look! Added:
* Engine check now calls `InitTranslationsBrowserState` on true success
    * `InitTranslationsBrowserState` calls `requestSupportedLanguages` (and will call other items in the future)

Thanks for mentioning this! Early when working on this, using coroutines was giving me a lot of trouble leaking middleware context via `context.dispatch` and I didn't know why. 😵‍💫 (So much so that @jonalmeida added to the error message in #5051). I for one would love more documentation in this area, especially when we have access to the wiki style docs since it is very subtle.

Reply:
Do you mind if I go ahead and allow this to land as is? I immediately have a chance to refactor in #5726 or I might just land again on this bug or new bug.

Testing this patch between changes is very difficult with all of the Fenix dependencies considerations too and I keep hitting CI build intermittent failures on random components. (I can't wait to Phabracator, which allows patch chaining to guarantee landing together, so that Fenix part could have been moved out easier.)

Is this what you are thinking?:
* `InitAction` to call `InitTranslationsBrowserState` directly 
    * Move the engine call out of `InitAction` that is now calling `InitTranslationsBrowserState`
* `InitTranslationsBrowserState` will check the engine and call the remaining functions, all within the same `scope.launch`

Refactor:
`InitAction`:
```
scope.launch {
  context.store.dispatch(InitTranslationsBrowserState)
}

```

`InitTranslationsBrowserState`:
(I think there is a a chance we may need to call this from other places too for error recovery, so I assume a second `scope.launch` is correct?)
```
scope.launch {
   checkEngine();
     -> checkEngine will call finishTranslationsInit()
}

```

 `checkEngine()`
```
// Check engine
// onSuccess  and engine works is true
finishTranslationsInit()

```

`finishTranslationsInit`:

```
fetchLanguages()
fetchSettings()
etc.

```

Reply:
Filed [1882403](https://bugzilla.mozilla.org/show_bug.cgi?id=1882403) for refactor, so it gets a new review, less entangled, etc.

Reply:
Yep, that's what I was thinking! I don't mind if you land as is and refactor, though I'll leave the final call up to @gabrielluong.

One small note though, in the updating `InitAction` handling:
```
scope.launch {
  context.store.dispatch(InitTranslationsBrowserState)
}
```

The launch isn't needed here, because the `Store::dispatch` does it

Reply:
I think it's fine to land as-is and followup. I am still catching up on the conversation here.

Reply:
Thanks all for the reviews! I'll immediately do the follow-up tomorrow, so the context doesn't get lost too.

Reply:
>I felt that you did what I was originally imagining in the https://github.com/mozilla-mobile/firefox-android/pull/5680#discussion_r1503387910

Yeah, that is what I followed, I thought "In case that suggestion is followed" was pointing back to [comment 1](https://github.com/mozilla-mobile/firefox-android/pull/5680#discussion_r1503387910) and focused in on the second paragraph in [comment 4](https://github.com/mozilla-mobile/firefox-android/pull/5680#discussion_r1503456124). I somehow missed the first paragraph in comment 4 saying avoid.

I can see some benefit to having `InitTranslationsBrowserState` for re-initializing everything too if an error occurs. (It adds complexity, but now I'm wondering if it is justified?)

Open to whatever refactor, let me know what you all prefer! I'll tag you both on the refactor review too, so we can pick it up from there.

Reply:
Sorry for muddying the waters! To clarify, my suggestion would be:

1. InitAction observed by Middleware
2. Middleware dispatches InitTranslationBrowserState through `context.store.dispatch`
3. Middleware observes `InitTranslationBrowserState`
4. scope.launch
    1.  checkEngine()
    2. finishTranslationsInit()

I think that will run the lowest risk of performance impact to BrowserStore startup and also atomize the initialization of the translation state.